### PR TITLE
Custom groups

### PR DESCRIPTION
**Pull request by mkraft**
#### Summary

* Adds the APIs, app methods, and store methods to:
    * create and delete custom groups
    * add members to and remove members from custom groups 
* Adds the ability to authorize against a new *group* permissions scope (using `(*App).SessionHasPermissionToGroup`)
* Adds a new `custom_group_user` role
* Adds new permissions:
    * `create_custom_group` (global scope)
    * `manage_custom_group_members ` (group scope)
    * `edit_custom_group ` (group scope)
    * `delete_custom_group ` (group scope)
* Adds a new search option for retrieving users who are *not* members of a given group
* Made the `(Group).RemoteId` field nullable

Unrelated changes:
* Fixed all outstanding lint warnings (except for 1)

#### Related PRs

* https://github.com/mattermost/mattermost-api-reference/pull/664
* https://github.com/mattermost/enterprise/pull/1097
* https://github.com/mattermost/mattermost-webapp/pull/9434

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-40042

#### Release Note

```release-note
Adds the APIs and app, store, and authorization methods to create and delete custom groups and add remove custom group members.
```

### CODE REVIEW

File: store/sqlstore/group_store.go [hunk 1]
```diff
@@ -1643,3 +1773,117 @@ func (s *SqlGroupStore) countTableWithSelectAndWhere(selectStr, tableName string
...
 	return count, nil
 }
+
+func (s *SqlGroupStore) UpsertMembers(groupID string, userIDs []string) ([]*model.GroupMember, error) {
+	members, query, args, err := s.buildUpsertMembersQuery(groupID, userIDs)
+	if err != nil {
+		return nil, err
+	}
+
+	if _, err = s.GetMaster().Exec(query, args...); err != nil {
```
Comment (mkraft) [inline store/sqlstore/group_store.go:1783]: Switch to sqlx

File: store/sqlstore/group_store.go [hunk 2]
```diff
@@ -1643,3 +1773,117 @@ func (s *SqlGroupStore) countTableWithSelectAndWhere(selectStr, tableName string
...
+			sq.Eq{"GroupId": groupID},
+			sq.Eq{"UserId": userIDs},
+			sq.Eq{"DeleteAt": 0},
+		}).
+		ToSql()
+	if err != nil {
+		return
+	}
+
+	_, err = s.GetReplica().Select(&members, membersSelectQuery, membersSelectArgs...)
```
Comment (mkraft) [inline store/sqlstore/group_store.go:1860]: sqlx

File: store/sqlstore/group_store.go [hunk 3]
```diff
@@ -1643,3 +1773,117 @@ func (s *SqlGroupStore) countTableWithSelectAndWhere(selectStr, tableName string
...
+	return
+}
+
+func (s *SqlGroupStore) DeleteMembers(groupID string, userIDs []string) ([]*model.GroupMember, error) {
+	members, query, args, err := s.buildDeleteMembersQuery(groupID, userIDs)
+	if err != nil {
+		return nil, err
+	}
+
+	if _, err = s.GetMaster().Exec(query, args...); err != nil {
```
Comment (mkraft) [inline store/sqlstore/group_store.go:1840]: sqlx

File: store/sqlstore/group_store.go [hunk 4]
```diff
@@ -1643,3 +1773,117 @@ func (s *SqlGroupStore) countTableWithSelectAndWhere(selectStr, tableName string
...
+		return nil, errors.Wrap(err, "failed to save GroupMember")
+	}
+
+	return members, err
+}
+
+func (s *SqlGroupStore) buildUpsertMembersQuery(groupID string, userIDs []string) (members []*model.GroupMember, query string, args []interface{}, err error) {
+	var retrievedGroup *model.Group
+	// Check Group exists
+	if err = s.GetReplica().SelectOne(&retrievedGroup, "SELECT * FROM UserGroups WHERE Id = :Id", map[string]interface{}{"Id": groupID}); err != nil {
```
Comment (mkraft) [inline store/sqlstore/group_store.go:1797]: sqlx

File: model/license.go [hunk 5]
```diff
@@ -259,6 +261,11 @@ func (f *Features) SetDefaults() {
 	if f.RemoteClusterService == nil {
 		f.RemoteClusterService = NewBool(*f.FutureFeatures)
 	}
+
+	if f.CustomGroups == nil {
+		// f.CustomGroups = NewBool(false) // TODO: change to this
+		f.CustomGroups = NewBool(*f.FutureFeatures) // TODO: remove this, it's just for testing
```
Comment (mkraft) [inline model/license.go:267]: This must be updated prior to merge. It's temporarily like this to avoid needing to create a new E20 license file, for now.

File: model/config.go [hunk 6]
```diff
@@ -365,6 +365,7 @@ type ServiceSettings struct {
 	ThreadAutoFollow                                  *bool   `access:"experimental_features"`
 	CollapsedThreads                                  *string `access:"experimental_features"`
 	ManagedResourcePaths                              *string `access:"environment_web_server,write_restrictable,cloud_restrictable"`
+	EnableCustomGroups                                *bool   // telemetry: none
```
Comment (mkraft) [inline model/config.go]: This will require `access` if it's exposed in the system console.

File: store/sqlstore/group_store.go [hunk 7]
```diff
@@ -263,6 +393,26 @@ func (s *SqlGroupStore) Delete(groupID string) (*model.Group, error) {
...
+		From("GroupMembers").
+		Where(sq.Eq{"UserId": userID}).
+		Where(sq.Eq{"GroupId": groupID}).
+		Where(sq.Eq{"DeleteAt": 0}).
+		ToSql()
+	if err != nil {
+		return nil, errors.Wrap(err, "get_member_query")
+	}
+	var groupMember model.GroupMember
+	err = s.GetMasterX().Get(&groupMember, query, args...)
```
Comment (streamer45) [inline store/sqlstore/group_store.go:408]: Do we mean to query master?

File: store/sqlstore/group_store.go [hunk 8]
```diff
@@ -116,6 +116,136 @@ func (s *SqlGroupStore) Create(group *model.Group) (*model.Group, error) {
...
+	if err = txn.Get(&newGroup, groupGroupQuery, g.Id, 1, 0); err != nil {
+		return nil, err
+	}
+	if err = txn.Commit(); err != nil {
+		return nil, err
+	}
+	return newGroup.ToModel(), nil
+}
+
+func (s *SqlGroupStore) checkUsersExist(userIDs []string) error {
```
Comment (streamer45) [inline store/sqlstore/group_store.go:205]: I didn't see a unit test for this. Looks like a great candidate.

File: model/client4.go [hunk 9]
```diff
@@ -7082,6 +7112,37 @@ func (c *Client4) PatchGroup(groupID string, patch *GroupPatch) (*Group, *Respon
...
+	defer closeBody(r)
+	var g []*GroupMember
+	if jsonErr := json.NewDecoder(r.Body).Decode(&g); jsonErr != nil {
+		return nil, nil, NewAppError("UpsertGroupMembers", "api.unmarshal_error", nil, jsonErr.Error(), http.StatusInternalServerError)
+	}
+	return g, BuildResponse(r), nil
+}
+
+func (c *Client4) DeleteGroupMembers(groupID string, userIds *GroupModifyMembers) ([]*GroupMember, *Response, error) {
+	payload, _ := json.Marshal(userIds)
```
Comment (streamer45) [inline model/client4.go:7133]: We should probably handle the error like in the above method.

File: go.tools.mod [hunk 10]
```diff
@@ -7,7 +7,7 @@ require (
 	github.com/golang-migrate/migrate/v4 v4.14.1 // indirect
 	github.com/jstemmer/go-junit-report v0.9.1 // indirect
 	github.com/jteeuwen/go-bindata v3.0.7+incompatible // indirect
-	github.com/mattermost/mattermost-utilities/mmgotool v0.0.0-20210721133912-8b250bf4d0f6 // indirect
+	github.com/mattermost/mattermost-utilities/mmgotool v0.0.0-20211006080735-07b1b58b8b09 // indirect
```
Comment (streamer45) [inline go.tools.mod:10]: Did you mean to update this?

File: app/group.go [hunk 11]
```diff
@@ -637,3 +720,57 @@ func (a *App) UserIsInAdminRoleGroup(userID, syncableID string, syncableType mod
...
+
+	for _, groupMember := range members {
+		a.publishGroupMemberEvent(model.WebsocketEventGroupMemberDelete, groupMember)
+	}
+
+	return members, nil
+}
+
+func (a *App) publishGroupMemberEvent(eventName string, groupMember *model.GroupMember) {
+	messageWs := model.NewWebSocketEvent(eventName, "", "", "", nil)
```
Comment (streamer45) [inline app/group.go:769]: Should we pass the userID? Otherwise this will possibly get broadcast to every connected user.

File: app/group.go [hunk 12]
```diff
@@ -574,11 +648,20 @@ func (a *App) GetGroupsByIDs(groupIDs []string) ([]*model.Group, *model.AppError
...
 func (a *App) ChannelMembersMinusGroupMembers(channelID string, groupIDs []string, page, perPage int) ([]*model.UserWithGroups, int64, *model.AppError) {
-	users, err := a.Srv().Store.Group().ChannelMembersMinusGroupMembers(channelID, groupIDs, page, perPage)
+	unsanitizedUsers, err := a.Srv().Store.Group().ChannelMembersMinusGroupMembers(channelID, groupIDs, page, perPage)
 	if err != nil {
 		return nil, 0, model.NewAppError("ChannelMembersMinusGroupMembers", "app.select_error", nil, err.Error(), http.StatusInternalServerError)
 	}
 
+	var users []*model.UserWithGroups
+	for _, u := range unsanitizedUsers {
+		us := a.sanitizeProfiles([]*model.User{&u.User}, false)
```
Comment (streamer45) [inline app/group.go:658]: Any reason for not calling `SanitizeProfile` instead?

File: app/group.go [hunk 13]
```diff
@@ -94,10 +103,51 @@ func (a *App) CreateGroup(group *model.Group) (*model.Group, *model.AppError) {
...
+		case errors.As(err, &invErr):
+			return nil, model.NewAppError("CreateGroupWithUserIds", "app.group.id.app_error", nil, invErr.Error(), http.StatusBadRequest)
+		case errors.As(err, &dupKey):
+			return nil, model.NewAppError("CreateGroup", "app.custom_group.unique_name", nil, dupKey.Error(), http.StatusBadRequest)
+		default:
+			return nil, model.NewAppError("CreateGroupWithUserIds", "app.insert_error", nil, err.Error(), http.StatusInternalServerError)
+		}
+	}
+
+	messageWs := model.NewWebSocketEvent(model.WebsocketEventReceivedGroup, "", "", "", nil)
```
Comment (streamer45) [inline app/group.go:124]: Who should this message reach to? Just the affected users or more?

File: app/group.go [hunk 14]
```diff
@@ -25,6 +25,15 @@ func (a *App) GetGroup(id string) (*model.Group, *model.AppError) {
 		}
 	}
 
+	if opts != nil && opts.IncludeMemberCount {
+		memberCount, err := a.Srv().Store.Group().GetMemberCount(id)
+		if err != nil {
+			return nil, model.NewAppError("GetGroup", "app.member_count", nil, err.Error(), http.StatusInternalServerError)
+		}
+		memberCountInt := int(memberCount)
+		group.MemberCount = &memberCountInt
```
Comment (streamer45) [inline app/group.go:34]: We could simply use `model.NewInt()`.

File: api4/handlers.go [hunk 15]
```diff
@@ -196,3 +199,13 @@ func (api *API) APILocal(h func(*Context, http.ResponseWriter, *http.Request)) h
 	}
 	return handler
 }
+
+func requireLicense(f handlerFunc) handlerFunc {
+	return func(c *Context, w http.ResponseWriter, r *http.Request) {
+		if c.App.Srv().License() == nil {
+			c.Err = model.NewAppError("", "api.license_error", nil, "", http.StatusNotImplemented)
```
Comment (streamer45) [inline api4/handlers.go:206]: Should the error just be a Forbidden? The server would be able to serve the request if the license was there. Same question for similar occurrences.

**Review by streamer45 — COMMENTED** [review]
Great work :muscle: Given the size, I've had a quick first pass and left some comments.

As a side question, should this go through Security review as well?

File: store/sqlstore/group_store.go [hunk 16]
```diff
@@ -116,6 +116,136 @@ func (s *SqlGroupStore) Create(group *model.Group) (*model.Group, error) {
...
+	if len(userIDs) > 0 {
+		usersSelectQuery, usersSelectArgs, err := s.getQueryBuilder().
+			Select("Id").
+			From("Users").
+			Where(sq.Eq{"Id": userIDs}).
+			ToSql()
+		if err != nil {
+			return err
+		}
```
[the remark this answers is not in the document]
Comment (mkraft) [reply store/sqlstore/group_store.go:214]: @BenCookie95 Was your intent to filter deleted users?

[the remark this answers is not in the document]
Comment (BenCookie95) [reply store/sqlstore/group_store.go:214]: We should probably filter out deleted users

File: store/sqlstore/group_store.go (hunk 8 above)
Comment (mkraft) [reply store/sqlstore/group_store.go:205]: Yeah. The behaviour is tested at https://github.com/mattermost/mattermost-server/pull/18839/files#diff-296cb9fe9469829f3e083acdb746280d12e303fada0491fa1225710004e5a780R345-R359. The one hesitation I have about exporting this to unit test it would be that it doesn't really belongs in the group store. It's more of a user store thing. Other options would be:
1. verify the user existence at another tier, [like the App](https://github.com/mattermost/mattermost-server/blob/custom_groups/app/group.go#L106)
2. add a foreign key constraint to group members to get the integrity check for free

I actually quite like option 2 because it's more resilient, more far-reaching, and avoids extra DB queries. That's something I think would be a good separate initiative. Thoughts?

File: app/group.go (hunk 13 above)
Comment (mkraft) [reply app/group.go:124]: I think this is the same reason as https://github.com/mattermost/mattermost-server/pull/18839#discussion_r770631824. @BenCookie95?

File: api4/handlers.go (hunk 15 above)
Comment (mkraft) [reply api4/handlers.go:206]: Yes I think you're right, 501 Not Implemented is the wrong code and 403 Forbidden seems more appropriate.

Comment (mkraft) [reply api4/handlers.go:206]: @streamer45 I just realized that while I agree, this is a breaking change because the existing license checks respond with StatusNotImplemented, so it'a bit of a bigger (and breaking) change.

Comment (streamer45) [reply api4/handlers.go:206]: Yes, I agree on 2 as well. Could we just file a ticket to address this on the next API version (if appropriate)?

File: api4/user.go [hunk 17]
```diff
@@ -967,6 +997,18 @@ func searchUsers(c *Context, w http.ResponseWriter, r *http.Request) {
 		}
 	}
 
+	if props.NotInGroupId != "" {
+		if c.App.Srv().License() == nil || !*c.App.Srv().License().Features.LDAPGroups {
```
Comment (BenCookie95) [inline api4/user.go:1001]: Should this do the `requireGroupAccess` check here?

File: api4/webhook_test.go [hunk 18]
```diff
@@ -719,7 +719,7 @@ func TestGetOutgoingWebhook(t *testing.T) {
 	CheckForbiddenStatus(t, resp)
 
 	th.TestForSystemAdminAndLocal(t, func(t *testing.T, client *model.Client4) {
-		nonExistentHook := &model.OutgoingWebhook{ChannelId: th.BasicChannel.Id}
+		nonExistentHook := &model.OutgoingWebhook{}
```
Comment (BenCookie95) [inline api4/webhook_test.go:722]: Why was this changed?

File: app/group.go (hunk 13 above)
Comment (BenCookie95) [reply app/group.go:123]: I actually didn't need the websocket for creating a group because we make a network request when we're typing to search for groups

File: app/group.go (hunk 11 above)
Comment (BenCookie95) [reply app/group.go:758]: For this I only wanted it for the user that is impacted, so we should just pass the userId. I also only need it for when deleting a user from a group and not when adding

**Review by BenCookie95 — CHANGES_REQUESTED** [review]

File: app/group.go (hunk 11 above)
Comment (mkraft) [reply app/group.go:758]: Ok, I'll make the change.

File: app/group.go (hunk 13 above)
Comment (mkraft) [reply app/group.go:123]: Ok, I can remove this one until it ever becomes necessary.

File: api4/webhook_test.go (hunk 18 above)
Comment (mkraft) [reply api4/webhook_test.go:722]: It was one of the linting changes I mentioned in the PR description. If I were to do it again I'd split them into a separate PR for ease of reviewing, but anyhow.

**Comment by mkraft** [comment]
> Great work 💪 Given the size, I've had a quick first pass and left some comments.
> 
> As a side question, should this go through Security review as well?

Thanks, yes it will go through a security review.

File: web/params.go [hunk 19]
```diff
@@ -355,6 +357,29 @@ func ParamsFromRequest(r *http.Request) *Params {
...
+	if val := query.Get("group_source"); val != "" {
+		switch val {
+		case "custom":
+			params.GroupSource = model.GroupSourceCustom
+		case "ldap":
+			params.GroupSource = model.GroupSourceLdap
+		default:
+			params.GroupSource = model.GroupSourceLdap
+		}
+	}
```
Comment (streamer45) [inline web/params.go:379]: Is the purpose of these to do some early validation or just setting the request param? I don't see them used anywhere.

File: app/group.go [hunk 20]
```diff
@@ -515,9 +568,13 @@ func (a *App) TeamMembersMinusGroupMembers(teamID string, groupIDs []string, pag
...
 	}
 
+	for _, u := range users {
+		a.SanitizeProfile(&u.User, false)
+	}
+
 	// parse all group ids of all users
 	allUsersGroupIDMap := map[string]bool{}
-	for _, user := range users {
+	for _, user := range users { //nolint
```
[the remark this answers is not in the document]
Comment (mkraft) [reply app/group.go]: I can take those out now.

File: api4/group.go [hunk 21]
```diff
@@ -108,36 +129,102 @@ func getGroup(c *Context, w http.ResponseWriter, r *http.Request) {
...
+	}
+
+	var groupPatch model.GroupPatch
+	if jsonErr := json.NewDecoder(r.Body).Decode(&groupPatch); jsonErr != nil {
+		c.SetInvalidParam("group")
+		return
+	}
+
+	if group.Source == model.GroupSourceCustom && groupPatch.AllowReference != nil && !*groupPatch.AllowReference {
+		c.Err = model.NewAppError("Api4.patchGroup", "api.custom_groups.must_be_referencable", nil, "", http.StatusBadRequest) // change to not implemented
```
[the remark this answers is not in the document]
Comment (mkraft) [reply api4/group.go]: Thanks, removed the old comment and fixed the typo.

File: web/params.go (hunk 19 above)
Comment (mkraft) [reply web/params.go]: It was added for use at https://github.com/mattermost/mattermost-server/pull/18839/files#diff-060eb667249b07159dcd89b520fa51777abf5bd355241d2e5c7e6bb58a93d367R852, but to me there's no need for both `props` and `query` retrieval, so I removed the props one. Also it was missing a test, which I added and that exposed that it was missing from the Go client4 too. Updated.

**Review by streamer45 — APPROVED** [review]
Looks great :+1:

**Comment by furqanmlk** [comment]
@BenCookie95 Can you please have a look at conflicts ?

**Comment by mkraft** [comment]
@BenCookie95 gentle reminder.

**Review by BenCookie95 — APPROVED** [review]

**Comment by cwarnermm** [comment]
Documented via https://github.com/mattermost/docs/pull/5406.
